### PR TITLE
feat: cypress web e2e

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

A minimal implementation of Cypress E2E with Typescript, designed to be a best practice template for common use cases, allowing other contributors to follow an existing pattern.

It provides mocks for a number of requests, though this will need to be expanded to support testing additional flows.

### Initial tests
- Render splash screen and show all supported wallets in the connect modal
- Can log in with locally stored Native wallet
- Can restore a Native wallet from the seed, and can then log in
- Account rows are rendered as expected (partial)
- Trades work as expected (partial)
- Send works as expected (partial)
- Receive works as expected (partial)

### Design decisions

- Implementation in line with [Cypress Best Practices](https://docs.cypress.io/guides/references/best-practices)
- Cypress TypeScript definitions are [isolated to avoid Jest and Cypress type conflicts](https://docs.cypress.io/guides/tooling/typescript-support#Clashing-types-with-Jest): separate `tsconfig.json` used for Cypress
- Uses [Cypress's official ESLint plugin](https://github.com/cypress-io/eslint-plugin-cypress)
- `chromeWebSecurity` is disabled due to a [Cypress limitation with Cross-origin iframes](https://docs.cypress.io/guides/guides/web-security#Cross-origin-iframes)
- Does not visit external sites (this prevents us from testing some flows, such as Portis login) via the UI, as this is considered a [Cypress anti-pattern](https://docs.cypress.io/guides/references/best-practices#Visiting-external-sites). One way to overcome this in the future is to interact with third-party servers programmatically - though this should be done sparingly, see point below.
- Minimises interaction with third-party servers, i.e. those that we do not control, where possible - see best practice [here](https://docs.cypress.io/guides/references/best-practices#3rd-party-servers). Acceptable exceptions include testing critical flows.
- [Prefers stubbed responses from ShapeShift-controlled servers](https://docs.cypress.io/guides/references/best-practices#3rd-party-servers) over full E2E tests. Full E2E is only recommended for testing critical paths. It also helps us overcome the issues inherent with testing _real_ addresses - unless we run a separate node, we cannot ensure the wallet properties remain static.
- Use Typescript `factories` instead of the Cypress default of JSON `fixtures` to give us type safety, reduce test data coupling, and minimize duplicate code

### Known issues and future improvements

- We can't currently mock data provided via websockets (which includes transactions). I think I've found every way that doesn't work. We could either put more resources into getting this working or use a real wallet.
- We don't yet have a programmatic way of getting to a "logged in" / "wallet connected" state - setting up `indexedDB` is not enough, as we also require React state to be set for the wallet (not Cypress's forte). Another factor likely to block this is the current, known, behavor of `web` where it resets wallet connection state on refresh.

### Best practices we are  currently breaking

- "Set state programmatically, don't use the UI to create state."

I suggest we write up a bounty to resolve this.

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?
- [ ] Do all new and existing tests pass? Does the linter pass?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

https://github.com/shapeshift/web/issues/596

## Testing

Please outline all testing steps:

Run using GUI: `yarn test:cypress`
Run headless: `test:cypress:headless`

## Screenshots (if applicable)
Screen Shot 2022-01-26 at 5 39 07 am

Screen Shot 2022-02-04 at 5 37 15 am

### CODE REVIEW

**Comment by rbeer** [comment]
Added the `data-test` attribute removal in my fork:
rbeer/shapeshift-web@42e55a2

File: src/context/WalletProvider/NativeWallet/components/NativeCreate.tsx [hunk 1]
```diff
@@ -19,6 +19,13 @@ import { Text } from 'components/Text'
 
 import { NativeSetupProps } from '../types'
 
+export const getVault = async (): Promise<Vault> => {
```
Comment (cjthompson) [inline src/context/WalletProvider/NativeWallet/components/NativeCreate.tsx:22]: Why does this need to be exported?

**Review by cjthompson — APPROVED** [review]

File: src/context/WalletProvider/NativeWallet/components/NativeCreate.tsx (hunk 1 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/NativeWallet/components/NativeCreate.tsx]: It's used in `cypress/helpers.ts`. That said, that helper it is used in is no longer needed after a recent refactor, so this has been addressed in https://github.com/shapeshift/web/pull/849/commits/857b2a63e5a58f59cd6bcc0b5f825835dc3c87c3.

**Review by 0xdef1cafe — APPROVED** [review]
beastly

File: cypress/plugins/index.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+require('dotenv').config()
+
+const execa = require('execa')
+
+const findBrave = (): Cypress.Browser => {
+  // the path is hard-coded for simplicity
+  const browserPath = '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply cypress/plugins/index.ts:7]: Yes, definitely something we need to fix ASAP.

MacOS does have `which`, but when Brave is installed `which brave-browser` doesn't work like it does on Linux.

**Comment by 0xApotheosis** [comment]
Quality feedback, thank you @rbeer.

PR with fixes incoming.

**Comment by 0xApotheosis** [comment]
For posterity: PR adding Brave support for Linux, and addressing README comments is here: https://github.com/shapeshift/web/pull/954

**Pull request merged**